### PR TITLE
drivers: versal: PM, MBOX and NVM

### PR DESCRIPTION
**Pull request by ldts**
The PLM firmware controls several subsystems in the Versal ACAD
platform. The NVM driver provides access to the PLM NVM subsystem interface via the MBOX interface.

### CODE REVIEW

**Comment by ldts** [comment]
going to post a simpler version
the PMC can do non-word aligned accesses to/from TEE memory.

**Comment by ldts** [comment]
updated

File: core/drivers/versal_pm.c [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+	thread_smccc(&args);
+
+	if (payload) {
+		reg_pair_from_64(args.a0, &payload->data[1], &payload->data[0]);
+		reg_pair_from_64(args.a1, &payload->data[3], &payload->data[2]);
+		reg_pair_from_64(args.a2, &payload->data[5], &payload->data[4]);
+	}
+
+	if (IS_ENABLED(CFG_VERSAL_TRACE_PLM))
+		mdelay(1000);
```
Comment (jforissier) [inline core/drivers/versal_pm.c:87]: What's the purpose of this delay?

File: core/include/drivers/versal_nvm.h [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+	EFUSE_OFFCHIP_REVOKE_ID_1,
+	EFUSE_OFFCHIP_REVOKE_ID_2,
+	EFUSE_OFFCHIP_REVOKE_ID_3,
+	EFUSE_OFFCHIP_REVOKE_ID_4,
+	EFUSE_OFFCHIP_REVOKE_ID_5,
+	EFUSE_OFFCHIP_REVOKE_ID_6,
+	EFUSE_OFFCHIP_REVOKE_ID_7
+};
+
+#define __aligned_efuse			__aligned(CACHELINE_LEN)
```
Comment (jforissier) [inline core/include/drivers/versal_nvm.h:56]: Is this expected to be used by the callers of the functions below? If not, please move it to `versal_nvm.c`.

**Review by jforissier — COMMENTED** [review]
- For commit "drivers: versal_pm: VERSAL ACAD PM driver":

  > Driver to communicate with TF-A executing at EL2

  Do you mean EL3? Or S-EL2 perhaps? Same in description for commit "drivers: versal_mbox: Versal ACAD mailbox driver".

  `Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

- For commits:
  "Retrieve the SoC revision from TF-A and show it in the banner"
  "plat-versal: define the cache line length"

  `Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

- For commit " drivers: versal_mbox: Versal ACAD mailbox driver":

  `Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

- For commit " drivers: versal_nvm: Versal ACAD non-volatile memory driver":
  One comment, then:

  `Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

File: core/drivers/versal_pm.c (hunk 1 above)
Comment (ldts) [reply core/drivers/versal_pm.c:87]: the PLM software running in the PMC uses the same UART than OP-TEE so this prevents the console from being scrambled when the PLM debug is enabled (just so we can debug the responses to OP-TEE requests...it was useful for me but I am ok to remove it if you think it makes little sense)

Comment (jforissier) [reply core/drivers/versal_pm.c:87]: That's OK but please add a comment here.

File: core/drivers/versal_nvm.c [hunk 3]
```diff
@@ -0,0 +1,496 @@
...
+};
+
+static TEE_Result prepare_cmd(struct ipi_cmd *cmd, enum versal_nvm_api_id efuse,
+			      struct ipi_buf *ibufs, struct cmd_args *arg)
+{
+	size_t i = 0;
+
+	cmd->data[i++] = NVM_API_ID(efuse);
+	for (i = 1; i < arg->len + 1; i++)
+		cmd->data[i] = arg->data[i];
```
Comment (ldts) [inline core/drivers/versal_nvm.c:257]: need to fix this (should be arg->data[i - 1])

**Review by etienne-lms — COMMENTED** [review]
minor comments only

**Comment by ldts** [comment]
going to close this PR - I ll send all the Versal drivers in a single PR in a couple of weeks

**Pull request closed without merging**